operators/pkg/tenant-controller/tenant_controller.go

```diff
@@ -307,19 +383,73 @@ func (r *TenantReconciler) checkValidWorkspaces(ctx context.Context, tn *crownla
 	return tenantExistingWorkspaces, workspaces, err
 }
 
+
+// Code modified from https://github.com/adracus/controller-runtime/commit/88fa42f9931939de0fef9cd6c118c531586d9153
+// and https://github.com/kubernetes-sigs/controller-runtime/blob/v0.13.1/pkg/controller/controllerutil/controllerutil.go#L195
+
+// mutate wraps a MutateFn and applies validation to its result.
+func mutate(f MutateFn, key client.ObjectKey, obj client.Object) error {
+	if err := f(); err != nil {
+		return err
+	}
+	if newKey := client.ObjectKeyFromObject(obj); key != newKey {
+		return fmt.Errorf("MutateFn cannot mutate object name and/or object namespace")
+	}
+	return nil
+}
+
+// MutateFn is a function which mutates the existing object into its desired state.
+type MutateFn func() error
+
+// DeleteIfExists deletes the given object in the Kubernetes
+// cluster if it exists.
+
+// It returns the executed operation and an error.
+func DeleteIfExists(ctx context.Context, c client.Client, obj client.Object, f MutateFn) (OperationResult string, err error) {
+	if err := c.Delete(ctx, obj); err != nil {
+		if apierrors.IsNotFound(err) {
+			return "unchanged", nil
+		}
+		return "unchanged", err
+	}
+	return "deleted", nil
+}
+
+
+// deleteClusterNamespace deletes the namespace for the tenant, if it fails then it returns an error
+func (r *TenantReconciler) deleteClusterNamespace(ctx context.Context, tn *crownlabsv1alpha2.Tenant, nsName string) (nsOk bool, err error) {
+	ns := v1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: nsName}}
+
+	_, nsErr := DeleteIfExists(ctx, r.Client, &ns, func() error {
+		r.updateTnNamespace(&ns, tn.Name)
+		return ctrl.SetControllerReference(tn, &ns, r.Scheme)
+	})
+
+	if nsErr != nil {
+		klog.Errorf("Error when deleting namespace of tenant %s -> %s", tn.Name, nsErr)
+		return false, nsErr
+	}
+
+	return true, nsErr
```

Comment:
You should use the `utils.EnforceObjectAbsence` function already leveraged e.g., [here](https://github.com/netgroup-polito/CrownLabs/blob/master/operators/pkg/tenant-controller/sandbox.go#L114)

Reply:
Resolved in commit [101d25b](https://github.com/netgroup-polito/CrownLabs/pull/821/commits/101d25bbfbfb4f9b9fe3301fb1814d67f416f3e9)